libraries/AP_Proximity/AP_Proximity_LightWareSerial.cpp

```diff
@@ -0,0 +1,153 @@
+        }
+    }
+
+    // checksum
+    _uart->write(LOWBYTE(crc));
+    _uart->write(HIGHBYTE(crc));
+}
+
+// process one byte received on serial port
+// returns true if a complete message has been received
+// state is stored in _msg structure
+bool AP_Proximity_LightWareSerial::parse_byte(uint8_t b)
+{
+    // check that payload buffer is large enough
+    static_assert(ARRAY_SIZE(_msg.payload) == PROXIMITY_LIGHTWARE_PAYLOAD_LEN_MAX, "AP_Proximity_LightWareSerial: check _msg.payload array size");
+
+    // process byte depending upon current state
+    switch (_msg.state) {
+
+    case ParseState::HEADER:
+        if (b == PROXIMITY_LIGHTWARE_HEADER) {
+            _msg.crc_expected = crc_xmodem_update(0, b);
+            _msg.state = ParseState::FLAGS_L;
+        }
+        break;
+
+    case ParseState::FLAGS_L:
+        _msg.flags_low = b;
+        _msg.crc_expected = crc_xmodem_update(_msg.crc_expected, b);
+        _msg.state = ParseState::FLAGS_H;
+        break;
+
+    case ParseState::FLAGS_H:
+        _msg.flags_high = b;
+        _msg.crc_expected = crc_xmodem_update(_msg.crc_expected, b);
+        _msg.payload_len = UINT16_VALUE(_msg.flags_high, _msg.flags_low) >> 6;
+        if ((_msg.payload_len == 0) || (_msg.payload_len > PROXIMITY_LIGHTWARE_PAYLOAD_LEN_MAX)) {
+            // invalid payload length, abandon message
+            _msg.state = ParseState::HEADER;
+        } else {
+            _msg.state = ParseState::MSG_ID;
+        }
+        break;
+
+    case ParseState::MSG_ID:
+        _msg.msgid = b;
+        _msg.crc_expected = crc_xmodem_update(_msg.crc_expected, b);
+        if (_msg.payload_len > 1) {
+            _msg.state = ParseState::PAYLOAD;
+        } else {
+            _msg.state = ParseState::CRC_L;
+        }
+        _msg.payload_recv = 0;
+        break;
+
+    case ParseState::PAYLOAD:
+        if (_msg.payload_recv < (_msg.payload_len - 1)) {
+            _msg.payload[_msg.payload_recv] = b;
+            _msg.payload_recv++;
+            _msg.crc_expected = crc_xmodem_update(_msg.crc_expected, b);
+        }
+        if (_msg.payload_recv >= (_msg.payload_len - 1)) {
+            _msg.state = ParseState::CRC_L;
+        }
+        break;
+
+    case ParseState::CRC_L:
+        _msg.crc_low = b;
+        _msg.state = ParseState::CRC_H;
+        break;
+
+    case ParseState::CRC_H:
+        _msg.state = ParseState::HEADER;
+        _msg.crc_high = b;
+        if (_msg.crc_expected == UINT16_VALUE(_msg.crc_high, _msg.crc_low)) {
+            return true;
+        }
+        break;
+    }
+
+    return false;
+}
+
+// convert buffer to uint32, uint16
+uint32_t AP_Proximity_LightWareSerial::buff_to_uint32(uint8_t b0, uint8_t b1, uint8_t b2, uint8_t b3) const
+{
+    uint32_t leval = (uint32_t)b0 | (uint32_t)b1 << 8 | (uint32_t)b2 << 16 | (uint32_t)b3 << 24;
+    return leval;
+}
+
+uint16_t AP_Proximity_LightWareSerial::buff_to_uint16(uint8_t b0, uint8_t b1) const
```

Comment:
We have 

```
#define UINT16_VALUE(hbyte, lbyte) (static_cast<uint16_t>((hbyte<<8)|lbyte))
```

in-tree now.  We should probably use it.

I wonder if other placed would benefit from a `UINT32_VALUE`.

Reply:
OK, removed buff_to_uint16 and replaced it with UINT16_VALUE.

Happy to create a new UINT32_VALUE but I wonder what we would call the bytes.  b0, b1, b2, b3 as I have here?  "low" and "high" isn't enough because we will have four arguments.

Reply:
I've created a UINT32_VALUE with arguments b3, b2, b1, b0.